### ISSUE TITLE
Proposal: ipython widget

### ISSUE THREAD

**Issue by chmp** [vega/ipyvega#93]
Having only recently discovered this great library, I experimented a bit with integrating altair / vega with ipywidgets and was wondering whether you would be interested in a contribution? In case you are, I would be happy to work on a PR.

You can see the current code [here](https://github.com/chmp/misc-exp/tree/master/vegawidget). It can be used as:

```python
import altair as alt
from vegawidget import VegaWidget

data = alt.InlineData([
    dict(c=0, v=1, t=1),
    dict(c=1, v=2, t=2),
    dict(c=2, v=3, t=3),
    dict(c=0, v=4, t=4),
    dict(c=1, v=5, t=5),
    dict(c=0, v=6, t=6),
    dict(c=0, v=7, t=7),
    dict(c=1, v=8, t=8),
    dict(c=2, v=9, t=9),
])

chart = alt.Chart(data).mark_bar().encode(x='c:O', y='sum(v):Q')

# update the spec to include a data name
# TODO: once on PyPI use InlineData(name='...')
spec = chart.to_dict()
spec['data']['name'] = 'table'

widget = VegaWidget(spec=spec)
display(widget)
```

In addition to the existing display system, it allows 

- to put altair charts in a larger UI with other widgets,
- to change the chart without re-displaying the widget, and 
- to dynamically update the data via changesets.

```python
# update the plot dynamically
for t in range(10, 20):
    value = dict(c=random.choice([0, 1, 2]), v=random.randint(1, 10), t=t)
    widget.update('table', remove=f'datum.t < {t - 5}', insert=[value])
    time.sleep(1)
```

**Comment by domoritz** [comment]
uhh, this is pretty sweet. It would be great if we could support Vega Signals as well and once we implement https://github.com/vega/vega-lite/issues/2790, this would be super useful for Vega-Lite and Altair (cc @arvind). 

@chmp We are always open to contributions! Do you want to send a PR and take it from there?

**Comment by chmp** [comment]
Sound good. I'll have a look how to integrate everything and will open a PR soonish.

**Comment by chmp** [comment]
@domoritz I will just this issue as hub for the whole widget issue... 

I think I found out what went wrong (and OMG I do hate the state of frontend dev) ... What I came up with is the following:

1. The module `@jupyter-widgets/base` is not an actual module. It is merely an alias to `nbextensions/jupyter-js-widgets/extension`.
2. This alias is defined in the widget extension [here](https://github.com/jupyter-widgets/ipywidgets/blob/master/widgetsnbextension/src/extension.js). After the widget extension has been loaded this alias is known to requirejs and our code that relied on this alias works as intended
4. For some reason, it seems that the load order of extensions depends on whether an extension was installed in development mode or as a normal package. I guess we both tested with a package installed in development mode and not the normal installation process. It seems in this case the widget extension was already loaded and we both did not observe the error.

In any case, while looking at the [widget-cookiecutter](https://github.com/jupyter-widgets/widget-cookiecutter) project, I realized that it is possible to ship multiple javascript files. Here we could ship two javascript files one containing the widget code, one containing the main vega extension. The advantages would be:

- The main vega extension does need to know anything about the widget (and hence will not break, whathever the widget does)
- The widget code is only loaded when requested by the ipython widget library. Therefore we can skip all this brittle optional dependency code in JS

I will start to refactor the code and let you know how it goes.
